share/wake/lib/core/unsafe.wake

```diff
@@ -0,0 +1,142 @@
+
+# Turns an safe value into an unsafe value
+export def makeUnsafe (v: a): Unsafe a =
+    Unsafe v
+
+# maps a safe function over an unsafe value.
+#
+# Examples:
+# ```
+#   unsafeMap (_ + 10) (makeUnsafe 10) = makeUnsafe 20
+# ```
+export def unsafeMap (f: a => b) (x: Unsafe a): Unsafe b =
+    unsafeBind x (f _ | makeUnsafe)
+
+# Implements application of an unsafe function to an unsafe value.
+# This allows functions that do not take unsafe values to be applied
+# to unsafe values. The effect works well with currying so it
+# allows one to chain the applications. This especially useful
+# for constructing tuples from unsafe values
+#
+# Example:
+# ```
+#   tuple MyTuple =
+#      MyString: String
+#      MyInteger: Integer
+#      MyOtherString: String
+#
+#   makeUnsafe MyTuple
+#   | unsafeApp (makeUnsafe "foo")
+#   | unsafeApp (makeUnsafe 10)
+#   | unsafeApp (makeUnsafe "bar")
+# ```
+export def unsafeApp (f: Unsafe (a => b)) (x: Unsafe a): Unsafe b =
+    unsafeBind f (unsafeMap _ x)
+
+# Lets you turn an function returning an unsafe value into a unsafe function
+# that returns a safe value. This let's you use these functions more naturally
+# in tricky locations like inside of higher order functions with the wrong
+# type.
+#
+# Example:
+# ```
+#   def unsafeFiles = ...
+#   def unsafeRunner = ...
+#   (unsafeSubsume unsafeFiles).unsafeBind (\files
+#   unsafeRunner.unsafeBind (\runner
+#     makePlan ...
+#     | setPlanFnOutputs (\_ files outputDir)
+#     | runJobWith runner
+#   ))
+# ```
+export def unsafeSubsume (f: a => Unsafe b): Unsafe (a => b) =
+    makeUnsafe (\x (f x).getUnsafeValue)
+
+# Takes a function returning an unsafe value, and maps it over an unsafe
+# value, flattening out the double unsafe before returning. This is equivlent
+# to first using unsafeMap and then calling unsafeFlatten. This might not
+# sound that useful but it winds up being the primary way that unsafe
+# code is composed. It lets you chain multiple unsafe operations togethor
+# in such a way that you can access their pure values inside the mapping
+# function. unsafeFlatMap is unseful in a pipe sequence of unsafe operations.
+#
+# ```
+#   def myUnsafeRunner = ...
+#   unsafeFlatMap (\runner
+#     makePlan "unsafe" Nil "echo foo > foo.txt; echo bar > bar.txt"
+#     | setPlanFnOutputs runner
+#   ) unsafeRunner
+# ```
+export def unsafeFlatMap (f: a => Unsafe b) (x: Unsafe a): Unsafe b =
+    unsafeBind x f
+
+# Takes an unsafe value and pipes it through a function returning an
+# an unsafe value. This is equivlient to unsafeFlatMap with its
+# arguments reversed. This can be nice to have over unsafeFlatMap
+# for syntatic reasons. You can think of it as "binding" the unsafe
+# value to the argument of the function. This is particularlly
+# nice when a lambda is used for the function argument.
+#
+# Example:
+# ```
+#   x.unsafeBind (\x y.unsafeBind (\y f x y))
+# ```
+export def unsafeBind (x: Unsafe a) (f: a => Unsafe b): Unsafe b =
```

Comment:
I know you said in your latest PR comment that you believe that both the `unsafeFlatMap` and the `unsafeBind` argument orderings are useful, but I'm still a bit concerned about having two, more or less equivalent, APIs, since it's one extra thing for a user to have to know about and then have to remember which is which. In general, I feel that idiomatic wake tends to use more of `unsafeFlatMap`'s style. I haven't pored through the codebase for this, but my gut feeling is that we tend to use `|` for functions of arity greater than 1 whenever the function is like a "method". This allows us to mimic a fluent style that most other languages get with `.` chaining, but we use `|` instead. I think we tend to only use `.` for 1-arity functions, where it looks like an "attribute access".

I think your example in the comment above would look pretty similar with `| unsafeFlatMap`:

```wake
x | unsafeFlatMap (\x y | unsafeFlatMap (\y f x y))
```

I acknowledge that this is a subjective, aesthetic opinion of mine, so I don't think it's worth blocking over this if you feel strongly about keeping it.

Reply:
I'm happy with just unsafeFlatMap, I can reorder things using `|` as suggested